File: docs/calidad/tuberias-ia/integracion-adenda-ia-frame-y-catalogo-errores/s17-contrato-recorrido-conjunto/administracion/publicar.py
from pathlib import Path
import sys,json,base64,hashlib
sys.path.insert(0,str(Path(__file__).resolve().parent.parent/'s6-trazabilidad-total'))
from github_io import fetch,call
R=Path(__file__).resolve().parent;CP=R/'PUBLICACION.json'
deletes=json.loads((R/'ELIMINACIONES.json').read_text())
maps=json.loads((R/'MAPA.json').read_text());state=json.loads(CP.read_text()) if CP.exists() else {}
def save():CP.write_text(json.dumps(state,ensure_ascii=False,indent=2)+'\n')
def blob(b):return hashlib.sha1(b'blob '+str(len(b)).encode()+b'\0'+b).hexdigest()
for k,m in maps.items():
 b=json.loads((R/(k+'-base.json')).read_text());s=state.setdefault(k,dict(base=b,files={}));assert s['base']==b
 api='https://api.github.com/repos/'+b['repo']
 if not s.get('published'):assert fetch(api+'/git/ref/heads/'+b['branch'])['object']['sha']==b['head']
for k,m in maps.items():
 s=state[k];b=s['base'];api='https://api.github.com/repos/'+b['repo']
 if s.get('verified'):continue
 old=json.loads((R/(k+'-arbol.json')).read_text())
 for path in deletes.get(k,[]):
  assert path in old;s['files'][path]=None;save()
 for path,local in m.items():
  data=Path(local).read_bytes();h=blob(data)
  if old.get(path)==h:continue
  if s['files'].get(path)==h:continue
  got=call('create_blob',dict(repository_full_name=b['repo'],content=base64.b64encode(data).decode(),encoding='base64'))
  assert got['sha']==h;s['files'][path]=h;save()
 if not s.get('commit'):
  t=call('create_tree',dict(repository_full_name=b['repo'],base_tree_sha=b['tree'],tree_elements=[dict(path=p,mode='100644',type='blob',sha=h) for p,h in s['files'].items()]))
  s['tree']=t['sha'];save()
  c=call('create_commit',dict(repository_full_name=b['repo'],message='calidad: fijar contrato del recorrido documental conjunto; S17 / RETP-188',parent_sha=b['head'],tree_sha=s['tree']))
  s['commit']=c['sha'];save()
 if not s.get('published'):
  assert fetch(api+'/git/ref/heads/'+b['branch'])['object']['sha']==b['head']
  call('update_ref',dict(repository_full_name=b['repo'],branch_name=b['branch'],sha=s['commit'],force=False));s['published']=True;save()
 assert fetch(api+'/git/ref/heads/'+b['branch'])['object']['sha']==s['commit']
 t=fetch(api+'/git/trees/'+s['tree']+'?recursive=1');assert not t.get('truncated');after={x['path']:x['sha'] for x in t['tree'] if x['type']=='blob'}
 for p,local in m.items():assert after[p]==blob(Path(local).read_bytes())
 for p in deletes.get(k,[]):assert p not in after
 for p,h in old.items():
  if p not in s['files']:assert after.get(p)==h,p
 assert set(after)-set(old)<=set(s['files']);s['verified']=True;save()
 print(k,s['commit'],'VERIFICADO',flush=True)

File: docs/calidad/tuberias-ia/paridad-imagen-celula-matematica/estudio-nucleo-agentes/bis-c12/administracion/publicar.js

// Publicación y verificación directa, sin entorno local ni force push.
// Requiere F=c12files y cortes/árboles recuperados mediante github_fetch.
function unwrap(r){if(r.isError)throw Error(JSON.stringify(r));return r.structuredContent.result||r.structuredContent;}
async function fetchJSON(url){return JSON.parse(unwrap(await tools.mcp__codex_apps__github_fetch({url})).content);}
const B=load("B"),D=B+"/bis-c12",F=load("c12files");
for(const key of ["lab","repo"]){
 const repo=load(key),branch=key==="lab"?"lab/playground-sv-permanente":"main",base=load(key==="lab"?"labhead":"head");
 const cp=load("c12publish-"+key)||{blobs:{},base};
 const url="https://api.github.com/repos/"+repo;
 if(cp.done){const h=await fetchJSON(url+"/git/ref/heads/"+branch);if(h.object.sha!==cp.commit)throw Error("Cabecera posterior distinta");text({repo,commit:cp.commit,verified:true});continue;}
 const head=await fetchJSON(url+"/git/ref/heads/"+branch);
 if(head.object.sha!==base)throw Error("Corte remoto cambió: "+repo);
 const files={};
 for(const [p,s]of Object.entries(F)){
  if(key==="lab"&&p.includes("REGISTRO_EVOLUCION_TECNICA_PROYECTO."))continue;
  const q=key==="repo"?p:p.includes("/Inventario-sv/sucesos/")?"laboratorio/tareas-watson/sucesos-sv/"+p.split("/").pop():p.replace("docs/calidad/","laboratorio/tareas-watson/");
  files[q]=s;
 }
 const before=load(key+"tree").tree.filter(x=>x.type==="blob");
 for(const [p,s]of Object.entries(files)){
  const prior=before.find(x=>x.path===p);if(prior&&prior.mode!=="100644")throw Error("Modo inesperado");
  if(!cp.blobs[p]){const b=unwrap(await tools.mcp__codex_apps__github_create_blob({repository_full_name:repo,content:s,encoding:"utf-8"}));if(!b.sha)throw Error("Sin SHA");cp.blobs[p]=b.sha;store("c12publish-"+key,cp);}
 }
 if(!cp.tree){
  const t=unwrap(await tools.mcp__codex_apps__github_create_tree({repository_full_name:repo,base_tree_sha:load(key+"commitbase").tree.sha,tree_elements:Object.entries(cp.blobs).map(([path,sha])=>({path,sha,mode:"100644",type:"blob"}))}));
  if(!t.sha)throw Error("Árbol sin SHA");cp.tree=t.sha;store("c12publish-"+key,cp);
 }
 const actual=await fetchJSON(url+"/git/trees/"+cp.tree+"?recursive=1");
 if(actual.truncated)throw Error("Árbol truncado");
 const after=actual.tree.filter(x=>x.type==="blob"),map=Object.fromEntries(after.map(x=>[x.path,x]));
 for(const x of before)if(!Object.hasOwn(files,x.path)){if(!map[x.path]||map[x.path].sha!==x.sha||map[x.path].mode!==x.mode)throw Error("Cambio ajeno "+x.path);}
 for(const [p,h]of Object.entries(cp.blobs))if(map[p]?.sha!==h)throw Error("Blob distinto "+p);
 if(after.some(x=>!before.some(b=>b.path===x.path)&&!Object.hasOwn(files,x.path)))throw Error("Archivo ajeno añadido");
 if(!cp.commit){const c=unwrap(await tools.mcp__codex_apps__github_create_commit({repository_full_name:repo,parent_sha:base,tree_sha:cp.tree,message:"docs: BIS-C12 perfiles, documentación y construcción; RETP-215"}));if(!c.sha)throw Error("Commit sin SHA");cp.commit=c.sha;store("c12publish-"+key,cp);}
 // Cotejar literalmente cada archivo preparado contra su blob creado.
 for(const [p,h]of Object.entries(cp.blobs)){
  const got=unwrap(await tools.mcp__codex_apps__github_fetch({url:"https://github.com/"+repo+"/blob/"+cp.commit+"/"+p})).content;
  if(got!==files[p])throw Error("Contenido distinto "+p);
 }
 const current=await fetchJSON(url+"/git/ref/heads/"+branch);if(current.object.sha!==base)throw Error("Cabecera cambió antes de publicar");
 unwrap(await tools.mcp__codex_apps__github_update_ref({repository_full_name:repo,branch_name:branch,sha:cp.commit,force:false}));
 const final=await fetchJSON(url+"/git/ref/heads/"+branch);if(final.object.sha!==cp.commit)throw Error("Cabecera no confirmada");
 const commit=await fetchJSON(url+"/git/commits/"+cp.commit);if(commit.tree.sha!==cp.tree||commit.parents[0].sha!==base)throw Error("Commit no concordante");
 cp.done=true;cp.count=Object.keys(files).length;store("c12publish-"+key,cp);text({repo,commit:cp.commit,files:cp.count,verified:true});
}
const a=load("c12publish-repo"),b=load("c12publish-lab");
for(const [p,h]of Object.entries(a.blobs)){
 if(p.includes("REGISTRO_EVOLUCION_TECNICA_PROYECTO."))continue;
 const q=p.includes("/Inventario-sv/sucesos/")?"laboratorio/tareas-watson/sucesos-sv/"+p.split("/").pop():p.replace("docs/calidad/","laboratorio/tareas-watson/");
 if(b.blobs[q]!==h)throw Error("Espejo diferente "+p);
}
text("Publicación, contenido literal, continuidad de árboles y espejo verificados. Checkout local pendiente de sincronizar.");
